### ISSUE TITLE
Gradient of Dirichlet with small `alpha` element gives `nan`

### ISSUE THREAD

**Issue by hylkedonker** [google/jax#18129]
### Description

Hi,

Thanks for building this awesome library! I am still amazed that you can take derivatives of random variables like the Dirichlet distribution, while JAX takes care of the reparameterisations. 🤯
While playing around, I came across an edge case where taking the derivative of a `jax.random.dirichlet` sample gives a `nan`.

```python
import jax
import jax.numpy as jnp

def sum_square(a, key):
    y = jax.random.dirichlet(key, a)
    return jnp.sum(y**2)

alpha = jnp.array([[2.4587069e-37, 2.3883429e+00, 3.9916661e+00, 7.5055346e-02,
         4.4362344e-02, 3.1457476e-02]], dtype=jnp.float32)
grads = jax.grad(sum_square)(alpha, jax.random.key(0))
# Output grads:
# Array([[           nan, -1.3312605e-01,  4.2917587e-02, -4.4759928e-04,
#         -2.5203576e-07, -5.7851091e-33]], dtype=float32)
```

It is probably related to the fact that, in this case, `jax.random.dirichlet` samples an exact zero:
```python
jax.random.dirichlet(jax.random.key(0), alpha)
# Output:
# Array([[0.0000000e+00, 1.3876437e-01, 8.6123359e-01, 2.0013299e-06,
#        3.7550779e-10, 1.4884029e-36]], dtype=float32)
```

I suppose this is a bug, right?
Let me know if there is anything I can do to further clarify.

Kind regards,

Hylke

### What jax/jaxlib version are you using?

jax-0.4.18/jaxlib-0.4.18

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.11.2, Ubuntu 23.04

**Comment by jakevdp** [comment]
Hi - thanks for the report! I think this doesn't necessarily have to do with `dirichlet`, but rather with the fact that the function $f(x) = \sqrt{x^2}$ doesn't have well-defined gradients as $x\to 0$:
```python
>>> jax.grad(lambda x: jax.numpy.sqrt(x ** 2))(0.0)
Array(nan, dtype=float32, weak_type=True)
```
Given this, I think the example is working as expected, and if you want different behavior near `x=0` you'll need to use a different functional form. What do you think?

**Comment by hylkedonker** [comment]
Thanks for your response. 
I think I don't completely follow you. Where does `sqrt` enter in the equation? Is `jax.random.dirichlet` using `jax.numpy.sqrt` under the hood?  
Thanks in advance,
Hylke

**Comment by jakevdp** [comment]
Oh, sorry, I thought you were computing a norm. Let me look closer at this...

**Comment by jakevdp** [comment]
It looks like you're right that there's something in the `dirichlet` implementation that leads to `NaN` gradients for very small values of `alpha`. Here's a shorter repro:
```python
def f(a, key):
  return jax.random.dirichlet(key, a[None])[0]

key = jax.random.key(0)
alpha = 1E-37

jax.grad(f)(alpha, key)
# NaN
```
